File: 1023-CamelcaseMatching/1023-CamelcaseMatching.java
/*

1023. Camelcase Matching
Medium

A query word matches a given pattern if we can insert lowercase letters to the pattern word so that it equals the query. (We may insert each character at any position, and may insert 0 characters.)

Given a list of queries, and a pattern, return an answer list of booleans, where answer[i] is true if and only if queries[i] matches the pattern.

Example 1:

Input: queries = ["FooBar","FooBarTest","FootBall","FrameBuffer","ForceFeedBack"], pattern = "FB"
Output: [true,false,true,true,false]
Explanation: 
"FooBar" can be generated like this "F" + "oo" + "B" + "ar".
"FootBall" can be generated like this "F" + "oot" + "B" + "all".
"FrameBuffer" can be generated like this "F" + "rame" + "B" + "uffer".

Example 2:

Input: queries = ["FooBar","FooBarTest","FootBall","FrameBuffer","ForceFeedBack"], pattern = "FoBa"
Output: [true,false,true,false,false]
Explanation: 
"FooBar" can be generated like this "Fo" + "o" + "Ba" + "r".
"FootBall" can be generated like this "Fo" + "ot" + "Ba" + "ll".

Example 3:

Input: queries = ["FooBar","FooBarTest","FootBall","FrameBuffer","ForceFeedBack"], pattern = "FoBaT"
Output: [false,true,false,false,false]
Explanation: 
"FooBarTest" can be generated like this "Fo" + "o" + "Ba" + "r" + "T" + "est".

Note:

    1 <= queries.length <= 100
    1 <= queries[i].length <= 100
    1 <= pattern.length <= 100
    All strings consists only of lower and upper case English letters.

*/

class Solution {
    public List<Boolean> camelMatch(String[] queries, String pattern) {
        Tnode root = new Tnode();
        insertToTrie(root, pattern);
        List<Boolean> resList = new ArrayList<>();
        for(String q : queries)
            resList.add(checkPattern(root, q));
        return resList;
    }
    private boolean checkPattern(Tnode root, String w){
        Tnode temp = root;
        for(char ch : w.toCharArray()){
            if(temp.children.containsKey(ch)){
                temp = temp.children.get(ch);
                continue;
            }
            if(ch >= 'a' && ch <= 'z') continue;
            return false;
        }
        return temp.isEnd;
    }
    
    private void insertToTrie(Tnode root, String w){
        Tnode temp = root;
        for(char ch : w.toCharArray()){
            if(!temp.children.containsKey(ch)){
                temp.children.put(ch, new Tnode());
                temp.ch = ch;
            }
            temp = temp.children.get(ch);
        }
        temp.isEnd = true;
    }
}
class Tnode{
    Map<Character, Tnode> children;
    char ch;
    boolean isEnd;
    Tnode(){
        this.children = new HashMap<>();
        this.isEnd = false;
        this.ch = '0';
    }
}

File: 1023-CamelcaseMatching/1023-CamelcaseMatching.py
"""

1023. Camelcase Matching

Given an array of strings queries and a string pattern, return a boolean array answer where answer[i] is true if queries[i] matches pattern, and false otherwise.

A query word queries[i] matches pattern if you can insert lowercase English letters into the pattern so that it equals the query. You may insert a character at any position in pattern or you may choose not to insert any characters at all.

Example 1:

Input: queries = ["FooBar","FooBarTest","FootBall","FrameBuffer","ForceFeedBack"], pattern = "FB"
Output: [true,false,true,true,false]
Explanation: "FooBar" can be generated like this "F" + "oo" + "B" + "ar".
"FootBall" can be generated like this "F" + "oot" + "B" + "all".
"FrameBuffer" can be generated like this "F" + "rame" + "B" + "uffer".

Example 2:

Input: queries = ["FooBar","FooBarTest","FootBall","FrameBuffer","ForceFeedBack"], pattern = "FoBa"
Output: [true,false,true,false,false]
Explanation: "FooBar" can be generated like this "Fo" + "o" + "Ba" + "r".
"FootBall" can be generated like this "Fo" + "ot" + "Ba" + "ll".

Example 3:

Input: queries = ["FooBar","FooBarTest","FootBall","FrameBuffer","ForceFeedBack"], pattern = "FoBaT"
Output: [false,true,false,false,false]
Explanation: "FooBarTest" can be generated like this "Fo" + "o" + "Ba" + "r" + "T" + "est".

Constraints:

    1 <= pattern.length, queries.length <= 100
    1 <= queries[i].length <= 100
    queries[i] and pattern consist of English letters.

"""

class TNode:
    def __init__(self):
        self.children = {}
        self.isend = False

class Solution:
    def camelMatch(self, queries: List[str], pattern: str) -> List[bool]:
        res = [False] * len(queries)
        root = TNode()
        def insert(word):
            temp = root
            for ch in word:
                if ch not in temp.children: temp.children[ch] = TNode()
                temp = temp.children[ch]
            temp.isend = True
        insert(pattern)
        
        def check_pattern(word):
            temp = root
            for ch in word:
                if ch in temp.children:
                    temp = temp.children[ch]
                    continue
                if ch >= 'a' and ch <= 'z': continue
                return False
            return temp.isend
        for i, query in enumerate(queries): res[i] = check_pattern(query)
        return res
        
